File: src/main/java/org/cekpelunasan/configuration/PlaywrightBrowserPool.java
package org.cekpelunasan.configuration;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.impl.TargetClosedError;
import jakarta.annotation.PreDestroy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

/**
 * Mengelola kumpulan (pool) instance browser Playwright agar PDF bisa dirender
 * secara paralel tanpa membuat browser baru setiap saat.
 *
 * <h3>Thread-safety</h3>
 * <p>
 * Playwright <strong>tidak thread-safe</strong> — objek Playwright hanya boleh diakses
 * dari thread yang membuatnya. Untuk itu, setiap slot di pool memiliki
 * {@link ExecutorService} satu-thread tersendiri. Semua operasi Playwright
 * (termasuk pembuatan, penggunaan, dan penggantian) selalu berjalan di thread
 * khusus milik slot tersebut.
 * </p>
 *
 * <h3>Lifecycle sebuah slot</h3>
 * <ol>
 *   <li>Saat startup: {@code Playwright.create()} + {@code chromium().launch()} dijalankan
 *       di thread khusus slot tersebut.</li>
 *   <li>Saat {@link #withBrowser}: slot diambil dari antrian; koneksi dicek di thread khusus;
 *       jika mati, direinisialisasi sebelum operasi dijalankan.</li>
 *   <li>Setelah operasi selesai (sukses maupun gagal): slot dikembalikan ke antrian.</li>
 *   <li>Saat shutdown: semua browser dan Playwright ditutup di thread khusus masing-masing.</li>
 * </ol>
 */
@Component
public class PlaywrightBrowserPool {

	private static final Logger log = LoggerFactory.getLogger(PlaywrightBrowserPool.class);

	/** Jumlah slot browser yang disiapkan sekaligus di dalam pool. */
	private static final int POOL_SIZE = 3;

	/** Batas waktu tunggu saat semua browser sedang terpakai, dalam detik. */
	private static final int ACQUIRE_TIMEOUT_SECONDS = 30;

	/** Batas waktu untuk satu operasi Playwright (safety net), dalam detik. */
	private static final int OPERATION_TIMEOUT_SECONDS = 90;

	private final BrowserType.LaunchOptions launchOptions;
	private final BlockingQueue<BrowserSlot> pool = new ArrayBlockingQueue<>(POOL_SIZE);
	private final List<BrowserSlot> allSlots = new ArrayList<>();

	/**
	 * Operasi yang dijalankan menggunakan instance {@link Browser} dari pool.
	 */
	@FunctionalInterface
	public interface BrowserOperation<T> {
		T execute(Browser browser) throws Exception;
	}

	/**
	 * Membuat pool: tiap slot mendapat thread khusus dan pasangan (Playwright, Browser) tersendiri.
	 *
	 * @param playwrightLaunchOptions opsi peluncuran browser dari {@link PlaywrightConfiguration}
	 */
	public PlaywrightBrowserPool(BrowserType.LaunchOptions playwrightLaunchOptions) {
		this.launchOptions = playwrightLaunchOptions;
		for (int i = 0; i < POOL_SIZE; i++) {
			try {
				BrowserSlot slot = new BrowserSlot(i, launchOptions);
				pool.offer(slot);
				allSlots.add(slot);
			} catch (Exception e) {
				log.error("Gagal membuat browser slot #{} saat inisialisasi pool", i, e);
			}
		}
		log.info("Playwright browser pool diinisialisasi dengan {} slot", pool.size());
	}

	/**
	 * Mengeksekusi operasi browser menggunakan slot yang tersedia dari pool.
	 * <p>
	 * Semua operasi Playwright berjalan di thread khusus milik slot, memastikan
	 * thread-safety Playwright terpenuhi. Jika browser dalam slot terputus saat
	 * akan dipakai, browser direinisialisasi secara otomatis di thread yang sama
	 * sebelum operasi dijalankan.
	 * </p>
	 *
	 * @param operation operasi yang akan dijalankan menggunakan {@link Browser}
	 * @param <T>       tipe nilai yang dikembalikan oleh operasi
	 * @return hasil dari operasi
	 * @throws InterruptedException  jika thread diinterupsi saat menunggu slot tersedia
	 * @throws IllegalStateException jika tidak ada slot tersedia setelah timeout
	 * @throws Exception             jika operasi melempar exception
	 */
	public <T> T withBrowser(BrowserOperation<T> operation) throws Exception {
		BrowserSlot slot = pool.poll(ACQUIRE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
		if (slot == null) {
			throw new IllegalStateException(
				"Tidak ada browser tersedia di pool setelah " + ACQUIRE_TIMEOUT_SECONDS + " detik");
		}
		try {
			return slot.execute(operation, launchOptions);
		} finally {
			pool.offer(slot);
		}
	}

	/**
	 * Menutup semua browser dan Playwright saat aplikasi shutdown.
	 */
	@PreDestroy
	public void shutdown() {
		for (BrowserSlot slot : allSlots) {
			slot.close();
		}
		log.info("Playwright browser pool dimatikan");
	}

	// -------------------------------------------------------------------------
	// Inner class: BrowserSlot
	// -------------------------------------------------------------------------

	private static class BrowserSlot {

		private static final Logger log = LoggerFactory.getLogger(BrowserSlot.class);

		private final int id;
		private final ExecutorService executor;
		private Playwright playwright;
		private Browser browser;

		/**
		 * Membuat slot baru: inisialisasi Playwright dan Browser di thread khusus.
		 */
		BrowserSlot(int id, BrowserType.LaunchOptions options) throws Exception {
			this.id = id;
			this.executor = Executors.newSingleThreadExecutor(r -> {
				Thread t = new Thread(r, "playwright-browser-" + id);
				t.setDaemon(true);
				return t;
			});
			// Playwright dan Browser harus dibuat di thread yang sama yang akan memakainya
			executor.submit(() -> init(options)).get(30, TimeUnit.SECONDS);
		}

		int id() {
			return id;
		}

		/** Membuat pasangan baru (Playwright, Browser) di thread saat ini. */
		private void init(BrowserType.LaunchOptions options) {
			playwright = Playwright.create();
			browser = playwright.chromium().launch(options);
			log.debug("Browser slot {} diinisialisasi", id);
		}

		/** Menutup pasangan lama dan membuat yang baru, di thread saat ini. */
		private void reinit(BrowserType.LaunchOptions options) {
			try { browser.close(); } catch (Exception ignored) {}
			try { playwright.close(); } catch (Exception ignored) {}
			init(options);
			log.info("Browser slot {} direinisialisasi setelah disconnect/crash", id);
		}

		/**
		 * Menjalankan operasi di thread khusus slot ini.
		 * Jika browser terputus, direinisialisasi terlebih dahulu.
		 */
		<T> T execute(BrowserOperation<T> operation, BrowserType.LaunchOptions options) throws Exception {
			try {
				return executor.submit((Callable<T>) () -> {
					if (browser == null || !browser.isConnected()) {
						log.warn("Browser slot {} terputus, memperbarui...", id);
						reinit(options);
					}
					try {
						return operation.execute(browser);
					} catch (TargetClosedError tce) {
						log.warn("Browser slot {} crash saat operasi, reinit segera", id);
						reinit(options);
						throw tce;
					}
				}).get(OPERATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
			} catch (ExecutionException e) {
				Throwable cause = e.getCause();
				if (cause instanceof Exception ex) throw ex;
				throw new RuntimeException(cause);
			}
		}

		/** Menutup browser dan Playwright, lalu shutdown executor. */
		void close() {
			try {
				executor.submit(() -> {
					try { browser.close(); } catch (Exception ignored) {}
					try { playwright.close(); } catch (Exception ignored) {}
				}).get(10, TimeUnit.SECONDS);
			} catch (Exception ignored) {
			} finally {
				executor.shutdownNow();
			}
		}
	}
}

File: src/main/java/org/cekpelunasan/configuration/PlaywrightBrowserPool.kt
package org.cekpelunasan.configuration

import com.microsoft.playwright.Browser
import com.microsoft.playwright.BrowserType.LaunchOptions
import com.microsoft.playwright.Playwright
import jakarta.annotation.PreDestroy
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue

/**
 * Mengelola kumpulan (pool) instance browser Playwright agar scraping bisa berjalan
 * secara paralel tanpa membuat browser baru setiap saat.
 * 
 * 
 * Membuat browser baru itu mahal — perlu waktu dan memori. Dengan pool, kita siapkan
 * sejumlah browser di awal lalu pinjam-kembalikan saat dibutuhkan. Kalau semua browser
 * sedang terpakai, pemanggil akan menunggu (blocking) sampai ada yang dikembalikan.
 * 
 * 
 * 
 * Pool juga menangani kasus browser yang crash atau terputus: saat browser
 * dikembalikan dalam kondisi tidak tersambung, pool otomatis menggantinya dengan
 * instance browser yang segar.
 * 
 * 
 * 
 * Ukuran pool saat ini: {@value #POOL_SIZE} browser. Angka ini bisa disesuaikan
 * tergantung kebutuhan konkurensi dan kapasitas memori server.
 * 
 */
@Component
class PlaywrightBrowserPool(private val launchOptions: LaunchOptions?) {
    private val playwright: Playwright
    private val pool: BlockingQueue<Browser> = ArrayBlockingQueue<Browser>(POOL_SIZE)

    /**
     * Membuat instance Playwright, meluncurkan sejumlah browser sesuai [.POOL_SIZE],
     * lalu mengisi pool dengan browser-browser tersebut.
     * 
     * 
     * Constructor ini dipanggil otomatis oleh Spring saat aplikasi start. Jika ada browser
     * yang gagal diluncurkan, exception akan dilempar dan aplikasi tidak akan start.
     * 
     * 
     * @param launchOptions opsi peluncuran browser yang dikonfigurasi di [PlaywrightConfiguration]
     */
    init {
        this.playwright = Playwright.create()
        for (i in 0..<POOL_SIZE) {
            pool.offer(launchBrowser())
        }
        log.info("Playwright browser pool initialized with {} instances", POOL_SIZE)
    }

    /**
     * Mengambil satu browser dari pool untuk dipakai.
     * 
     * 
     * Jika semua browser sedang terpakai, method ini akan menunggu (blocking) sampai
     * ada browser yang dikembalikan ke pool lewat [.release].
     * Pastikan selalu memanggil `release()` setelah selesai, idealnya
     * di dalam blok `finally`.
     * 
     * 
     * @return instance [Browser] yang siap dipakai
     * @throws InterruptedException jika thread yang menunggu di-interrupt
     */
    @Throws(InterruptedException::class)
    fun acquire(): Browser {
        return pool.take()
    }

    /**
     * Mengembalikan browser yang sudah selesai dipakai ke pool.
     * 
     * 
     * Jika browser masih tersambung dengan baik, langsung dikembalikan ke antrian.
     * Tapi kalau browser crash atau terputus, pool akan meluncurkan browser baru
     * sebagai penggantinya sehingga jumlah browser di pool tetap terjaga.
     * 
     * 
     * @param browser instance browser yang ingin dikembalikan, boleh `null`
     * (akan diabaikan dan diganti dengan browser baru)
     */
    fun release(browser: Browser?) {
        if (browser != null && browser.isConnected()) {
            pool.offer(browser)
        } else {
            log.warn("Browser disconnected or null, replacing with fresh instance")
            synchronized(this) {
                try {
                    pool.offer(launchBrowser())
                } catch (e: Exception) {
                    log.error("Failed to replace crashed browser in pool", e)
                }
            }
        }
    }

    /**
     * Meluncurkan satu instance browser Chromium baru menggunakan opsi yang sudah dikonfigurasi.
     * 
     * 
     * Method ini disinkronisasi karena Playwright tidak thread-safe untuk operasi pembuatan browser.
     * 
     * 
     * @return instance [Browser] baru yang sudah siap dipakai
     */
    @Synchronized
    private fun launchBrowser(): Browser? {
        return playwright.chromium().launch(launchOptions)
    }

    /**
     * Menutup semua browser di pool dan menghentikan instance Playwright saat aplikasi shutdown.
     * 
     * 
     * Method ini dipanggil otomatis oleh Spring sebelum aplikasi berhenti. Semua error
     * saat menutup browser diabaikan agar proses shutdown tidak tersangkut.
     * 
     */
    @PreDestroy
    fun shutdown() {
        var b: Browser?
        while ((pool.poll().also { b = it }) != null) {
            try {
                b!!.close()
            } catch (ignored: Exception) {
            }
        }
        try {
            playwright.close()
        } catch (ignored: Exception) {
        }
        log.info("Playwright browser pool shut down")
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(PlaywrightBrowserPool::class.java)

        /** Jumlah browser yang disiapkan sekaligus di dalam pool.  */
        private const val POOL_SIZE = 3
    }
}
